cmd/cert-checker/main.go

```diff
@@ -0,0 +1,261 @@
+			Valid:    valid,
+			Problems: problems,
+		}
+		if !valid {
+			atomic.AddInt64(&c.issuedReport.BadCerts, 1)
+		} else {
+			atomic.AddInt64(&c.issuedReport.GoodCerts, 1)
+		}
+	}
+	wg.Done()
+}
+
+func (c *certChecker) checkCert(cert core.Certificate) (problems []string) {
+	// Check digests match
+	if cert.Digest != core.Fingerprint256(cert.DER) {
+		problems = append(problems, "Stored digest doesn't match certificate digest")
+	}
+
+	// Parse certificate
+	parsedCert, err := x509.ParseCertificate(cert.DER)
+	if err != nil {
+		problems = append(problems, fmt.Sprintf("Couldn't parse stored certificate: %s", err))
+	} else {
+		// Check stored serial is correct
+		if core.SerialToString(parsedCert.SerialNumber) != cert.Serial {
+			problems = append(problems, "Stored serial doesn't match certificate serial")
+		}
+		// Check we have the right expiration time
+		if !parsedCert.NotAfter.Equal(cert.Expires) {
+			problems = append(problems, "Stored expiration doesn't match certificate NotAfter")
+		}
+		// Check basic constraints are set
+		if !parsedCert.BasicConstraintsValid {
+			problems = append(problems, "Certificate doesn't have basic constraints set")
+		}
+		// Check the cert isn't able to sign other certificates
+		if parsedCert.IsCA {
+			problems = append(problems, "Certificate can sign other certificates")
+		}
+		// Check the cert has the correct validity period +/- an hour
+		if parsedCert.NotAfter.Sub(cert.Issued) > checkPeriod+time.Hour {
+			problems = append(problems, "Certificate has a validity period longer than 90 days")
+		} else if parsedCert.NotAfter.Sub(cert.Issued) < checkPeriod-time.Hour {
+			problems = append(problems, "Certificate has a validity period shorter than 90 days")
+		}
+		// Check that the PA is still willing to issue for each name in DNSNames + CommonName
+		for _, name := range append(parsedCert.DNSNames, parsedCert.Subject.CommonName) {
+			if err = c.pa.WillingToIssue(core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}); err != nil {
+				problems = append(problems, fmt.Sprintf("Policy Authority isn't willing to issue for %s: %s", name, err))
+			}
+		}
+		// Check the cert has the correct key usage extensions
+		if !core.CmpExtKeyUsageSlice(parsedCert.ExtKeyUsage, []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}) {
+			problems = append(problems, "Certificate has incorrect key usage extensions")
+		}
+	}
+	return problems
+}
+
+func main() {
+	app := cmd.NewAppShell("cert-checker", "Checks validity of certificates issued in the last 90 days")
+	app.App.Flags = append(app.App.Flags, cli.IntFlag{
+		Name:  "workers",
+		Value: runtime.NumCPU(),
+		Usage: "The number of concurrent workers used to process certificates",
+	}, cli.StringFlag{
+		Name:  "report-dir-path",
+		Usage: "The path to write a JSON report on the certificates checks to (if no path is provided the report will not be written out)",
+	}, cli.StringFlag{
+		Name:  "db-connect",
```

Comment:
for later clean-up, not this PR: I'm not sure why we're defining this as a flag here instead of using just the config file?

(If we do want this, it's worrisome that this defines a cli flag name instead of just a env var because use a flag puts secrets (like the db password) into the process name readable by others.)